modules/ephemeris/src/main/scala/gem/horizons/HorizonsClient.scala

```diff
@@ -31,13 +31,19 @@ import scala.concurrent.duration._
   */
 object HorizonsClient {
 
-  /** A stream that emits a single client that will be shut down automatically. An internal global
-    * mutex ensures that only one such stream exists at a time.
+  /** A global mutex for limiting concurrent connections to Horizons. */
+  private val mutex: Semaphore[IO] =
+    Semaphore[IO](1L)(implicitly, scala.concurrent.ExecutionContext.global).unsafeRunSync // note
+
+  /** A stream that emits a single client that will be shut down automatically. An global mutex
+    * ensures that only one such stream exists at a time.
     */
   val client: Stream[IO, Client[IO]] = {
-    val mutex  = Semaphore[IO](1L)(implicitly, scala.concurrent.ExecutionContext.global).unsafeRunSync // note
+    // By using the singleton type here we prove that the mutex is a constant value. This guards
+    // against someone accidentally refactoring it into a method.
+    val mutexʹ: mutex.type = mutex
```

Comment:
Can you explain a bit about `singleton types`? Is it enabled by TLS?

Reply:
Not TLC; it's been in Scala for a long time. The TL;DR is that a constant references have singleton types, so we're using that type explicitly to prove this.

Singleton types arise in a few places.

- The built-in literals have them, but that's not what we're doing here.

```scala
scala> val a = 42
a: Int = 42

scala> final val a = 42 // note the return type
a: Int(42) = 42

scala> val a: 42 = 42 // TLS only
a: 42 = 42

scala> val a: shapeless.Witness.`42`.T = 42
a: Int(42) = 42
```

- `object`s have singleton types.

```scala
scala> :t List
scala.collection.immutable.List.type

scala> val x: List.type = List
x: scala.collection.immutable.List.type = scala.collection.immutable.List$@323a351c

scala> x(1,2,3)
res0: List[Int] = List(1, 2, 3)
```

- *variables* have them. This is what we're doing here. Note that I didn't say *values*, although if `a.type =:= b.type` then certainly `a == b`, but it's much stronger than this. It means they're the same value. Note that singleton types won't normally be inferred.

```scala
scala> val a = "foo"
a: String = foo

scala> val b = "foo"
b: String = foo

scala> a == b
res2: Boolean = true

scala> implicitly[a.type =:= b.type]
<console>:14: error: Cannot prove that a.type =:= b.type.
       implicitly[a.type =:= b.type]
                 ^

scala> val b = a
b: String = foo

scala> implicitly[a.type =:= b.type]
<console>:14: error: Cannot prove that a.type =:= b.type.
       implicitly[a.type =:= b.type]
                 ^

scala> val b: a.type = a
b: a.type = foo

scala> implicitly[a.type =:= b.type]
res5: a.type =:= b.type = <function1>
```

- Note further than `var`s and `def`s don't have singleton types, which would be unsound because the values can change and thus their types would need to change too, which can't happen.

```scala
scala> var a = "foo"
a: String = foo

scala> val b: a.type = a
<console>:12: error: stable identifier required, but a.type found.
       val b: a.type = a
               ^

scala> def a = "foo"
a: String

scala> val b: a.type = a
<console>:12: error: stable identifier required, but a.type found.
       val b: a.type = a
               ^
```

There's a lot more to say about it but hopefully this answers your question.

Reply:
Great explanation. thanks!